dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp

```diff
@@ -214,12 +142,140 @@ class MockAsyncCallData : public IAsyncCallData
+        , log(Logger::get())
+        , recv_mem_tracker(MemoryTracker::create(1048576))
+    {
+        prepareMsgChannels();
+    }
+
+    ~MockExchangeReceiver()
+    {
+        waitAllConnectionDone();
+    }
+
+    void connectionDone(bool meet_error, const String & local_err_msg)
+    {
+        Int32 copy_connection = -1;
+        {
+            std::lock_guard<std::mutex> lock(mu);
+            if (meet_error)
+                err_msg = local_err_msg;
+            --live_connections;
+            copy_connection = live_connections;
+        }
+
+        if (meet_error || copy_connection == 0)
+        {
+            finishAllMsgChannels();
+            cv.notify_all();
+        }
+    }
+
+    void connectLocalTunnel(std::vector<MPPTunnelPtr> & tunnels)
+    {
+        if (static_cast<Int32>(tunnels.size()) != live_connections)
+            throw Exception("conn_num != tunnels.size()");
+
+        for (auto & tunnel : tunnels)
+        {
+            LocalRequestHandler local_request_handler(
+                reinterpret_cast<MemoryTracker *>(recv_mem_tracker->get()),
+                [this](bool meet_error, const String & err_msg) {
+                    this->connectionDone(meet_error, err_msg);
+                },
+                ReceiverChannelWriter(&msg_channels, "", log, &data_size_in_queue, ExchangeMode::Local));
+            tunnel->connectLocal(0, local_request_handler, false);
+        }
+    }
+
+    void run()
+    {
+        auto msg_channel_num = msg_channels.size();
+        size_t closed_num = 0;
+        while (true)
+        {
+            for (size_t i = 0; i < msg_channel_num; ++i)
+            {
+                std::shared_ptr<ReceivedMessage> recv_msg;
+                switch (msg_channels[i]->pop(recv_msg))
+                {
+                case DB::MPMCQueueResult::OK:
+                    received_msgs.push_back(recv_msg);
+                    break;
+                default:
+                    ++closed_num;
+                }
+            }
+
+            if (closed_num == msg_channel_num)
+                break;
+        };
+    }
+
+    std::vector<std::shared_ptr<ReceivedMessage>> & getReceivedMsgs() { return received_msgs; }
+
+    String getErrMsg() const { return err_msg; }
+
+private:
+    void waitAllConnectionDone()
+    {
+        std::unique_lock lock(mu);
+        auto pred = [&] {
+            return live_connections == 0;
+        };
+        cv.wait(lock, pred);
+    }
+
+    void finishAllMsgChannels()
+    {
+        for (auto & msg_channel : msg_channels)
+            msg_channel->finish();
+    }
+
+    void prepareMsgChannels()
+    {
+        if (enable_fine_grained_shuffle)
+            for (Int32 i = 0; i < output_stream_count; ++i)
+                msg_channels.push_back(std::make_shared<MPMCQueue<std::shared_ptr<ReceivedMessage>>>(10));
+        else
+            msg_channels.push_back(std::make_shared<MPMCQueue<std::shared_ptr<ReceivedMessage>>>(10));
```

Comment:
A tunnel is all that is needed here.
Don't copy too much of the ExchangeReceiver code, use only what is necessary to complete the tests.

Reply:
> A tunnel is all that is needed here. Don't copy too much of the ExchangeReceiver code, use only what is necessary to complete the tests.

okk